File: backend/filter.go
package main

import (
	"bytes"
	"runtime"
	"strings"
	"sync"
	"time"
)

// FilterParams describes a filtering request.
type FilterParams struct {
	Year      int      `json:"year"`
	StartTime string   `json:"startTime"` // "MM-DD HH:mm:ss" or ""
	EndTime   string   `json:"endTime"`   // "MM-DD HH:mm:ss" or ""
	Keywords  []string `json:"keywords"`  // all must match (AND), case-insensitive
	Level     string   `json:"level"`     // optional single level, "" = any
}

type LogEntry struct {
	LineNo   int64  `json:"lineNo"`
	Time     string `json:"time"`
	Level    string `json:"level"`
	Logger   string `json:"logger"`
	Thread   string `json:"thread"`
	Msg      string `json:"msg"`
	Unix     int64  `json:"unix"`
	HasTime  bool   `json:"hasTime"`
	Text     string `json:"text"`
	FileName string `json:"fileName"`
}

type FilterResult struct {
	Total   int   `json:"total"`
	Indices []int `json:"indices"`
}

func parseDT(year int, s string) (int64, bool) {
	s = strings.TrimSpace(s)
	if s == "" {
		return 0, false
	}
	// Accept "MM-DD HH:mm:ss" or "MM-DD HH:mm"
	layouts := []string{"01-02 15:04:05", "01-02 15:04", "15:04:05", "15:04", "2006-01-02 15:04:05", "2006-01-02 15:04"}
	for _, l := range layouts {
		if t, err := time.ParseInLocation(l, s, time.Local); err == nil {
			if strings.HasPrefix(l, "01-02") {
				t = time.Date(year, t.Month(), t.Day(), t.Hour(), t.Minute(), t.Second(), 0, time.Local)
			}
			return t.Unix(), true
		}
	}
	return 0, false
}

// filterLog matches records against params and returns matched indices.
// It runs in parallel across the available cores. Progress is reported via
// the supplied callback (bytes done / total bytes).
func filterLog(pl *ParsedLog, params FilterParams, progress func(done, total int64)) *FilterResult {
	var startUnix, endUnix int64
	hasStart, hasEnd := false, false
	if params.StartTime != "" {
		if u, ok := parseDT(params.Year, params.StartTime); ok {
			startUnix, hasStart = u, true
		}
	}
	if params.EndTime != "" {
		if u, ok := parseDT(params.Year, params.EndTime); ok {
			endUnix, hasEnd = u, true
		}
	}
	level := strings.ToUpper(strings.TrimSpace(params.Level))
	var kws [][]byte
	for _, k := range params.Keywords {
		k = strings.TrimSpace(k)
		if k != "" {
			kws = append(kws, bytes.ToLower([]byte(k)))
		}
	}

	records := pl.Records
	if len(records) == 0 {
		return &FilterResult{}
	}
	needText := len(kws) > 0

	nworkers := runtime.NumCPU()
	if nworkers > 8 {
		nworkers = 8
	}
	if nworkers < 1 {
		nworkers = 1
	}
	if nworkers > len(records) {
		nworkers = len(records)
	}

	chunkSize := (len(records) + nworkers - 1) / nworkers
	results := make([][]int, nworkers)
	var wg sync.WaitGroup

	for w := 0; w < nworkers; w++ {
		lo := w * chunkSize
		hi := lo + chunkSize
		if hi > len(records) {
			hi = len(records)
		}
		if lo >= hi {
			continue
		}
		wg.Add(1)
		go func(w, lo, hi int) {
			defer wg.Done()
			var matched []int
			// Read the whole contiguous byte range for this worker in one call.
			startOff := records[lo].Offset
			endOff := records[hi-1].Offset + records[hi-1].Length
			var buf []byte
			if needText {
				buf = make([]byte, endOff-startOff)
				_, _ = pl.fh.ReadAt(buf, startOff)
				buf = bytes.ToLower(buf)
			}
			for i := lo; i < hi; i++ {
				rec := records[i]
				if hasStart && rec.Unix < startUnix {
					continue
				}
				if hasEnd && rec.Unix > endUnix {
					continue
				}
				if level != "" && rec.Level != level {
					continue
				}
				if needText {
					text := buf[rec.Offset-startOff : rec.Offset-startOff+rec.Length]
					ok := true
					for _, kw := range kws {
						if !bytes.Contains(text, kw) {
							ok = false
							break
						}
					}
					if !ok {
						continue
					}
				}
				matched = append(matched, i)
			}
			results[w] = matched
		}(w, lo, hi)
	}
	wg.Wait()

	var all []int
	for w := range results {
		all = append(all, results[w]...)
	}
	res := &FilterResult{Indices: all, Total: len(all)}
	_ = progress
	return res
}

// getPage materializes a page of log entries from matched indices.
func getPage(pl *ParsedLog, indices []int, offset, limit int) []LogEntry {
	if offset < 0 {
		offset = 0
	}
	if limit <= 0 {
		limit = 100
	}
	start := offset
	end := offset + limit
	if start > len(indices) {
		start = len(indices)
	}
	if end > len(indices) {
		end = len(indices)
	}
	out := make([]LogEntry, 0, end-start)
	for _, idx := range indices[start:end] {
		rec := pl.Records[idx]
		out = append(out, LogEntry{
			LineNo:  rec.LineNo,
			Time:    rec.Time,
			Level:   rec.Level,
			Logger:  rec.Logger,
			HasTime: rec.HasTime,
			Text:    string(pl.readRecord(rec)),
		})
	}
	return out
}

File: frontend/src/dept/filter.js
// filter.js — 移植自 backend/filter.go
// 过滤、分页、CSV 导出构建。

import {parseLayout} from './date.js';
import {lineSpans, readRecord} from './parser.js';

// parseDT parses a filter time string into a unix timestamp.
export function parseDT(year, s) {
    s = (s || '').trim();
    if (s === '') return {unix: 0, ok: false};
    const layouts = ['01-02 15:04:05', '01-02 15:04', '15:04:05', '15:04', '2006-01-02 15:04:05', '2006-01-02 15:04'];
    for (const l of layouts) {
        const u = parseLayout(l, s, year);
        if (u !== 0) return {unix: u, ok: true};
    }
    return {unix: 0, ok: false};
}

// filterLog matches records against params and returns matched indices.
export function filterLog(pl, params) {
    let startUnix = 0, endUnix = 0, hasStart = false, hasEnd = false;
    if (params.startTime) {
        const r = parseDT(params.year, params.startTime);
        if (r.ok) { startUnix = r.unix; hasStart = true; }
    }
    if (params.endTime) {
        const r = parseDT(params.year, params.endTime);
        if (r.ok) { endUnix = r.unix; hasEnd = true; }
    }
    const level = (params.level || '').toUpperCase().trim();
    const kws = [];
    for (const k of (params.keywords || [])) {
        const kk = (k || '').trim();
        if (kk !== '') kws.push(kk.toLowerCase());
    }

    const records = pl.records;
    if (records.length === 0) return {total: 0, indices: []};
    const needText = kws.length > 0;

    const matched = [];
    const content = pl.content;
    for (let i = 0; i < records.length; i++) {
        const rec = records[i];
        if (hasStart && rec.unix < startUnix) continue;
        if (hasEnd && rec.unix > endUnix) continue;
        if (level !== '' && rec.level !== level) continue;
        if (needText) {
            const text = content.slice(rec.offset, rec.offset + rec.length).toLowerCase();
            let ok = true;
            for (const kw of kws) {
                if (!text.includes(kw)) { ok = false; break; }
            }
            if (!ok) continue;
        }
        matched.push(i);
    }
    return {total: matched.length, indices: matched};
}

// buildEntry materializes a LogEntry for a record.
export function buildEntry(pl, rec, fileName) {
    return {
        lineNo: rec.lineNo,
        time: rec.time,
        level: rec.level,
        logger: rec.logger,
        thread: rec.thread,
        msg: rec.msg,
        unix: rec.unix,
        hasTime: rec.hasTime,
        text: readRecord(pl, rec),
        fileName: fileName || '',
    };
}

// runFilter across multiple files, returning stats + a combined index.
export function runFilter(files, params) {
    if (files.length === 0) {
        return {total: 0, items: [], filesLoaded: 0, totalBytes: 0, physicalLines: 0, foldedLines: 0, message: '没有已加载的文件'};
    }
    const items = [];
    let totalBytes = 0, physicalLines = 0, foldedLines = 0;
    for (let fi = 0; fi < files.length; fi++) {
        const pl = files[fi];
        totalBytes += pl.size;
        const r = filterLog(pl, params);
        const spans = lineSpans(pl);
        for (const idx of r.indices) {
            items.push({file: fi, rec: idx});
            const sp = spans[idx];
            physicalLines += sp;
            foldedLines += sp - 1;
        }
    }
    return {total: items.length, items, filesLoaded: files.length, totalBytes, physicalLines, foldedLines};
}

// getPage materializes a page of entries from a combined result.
export function getPage(files, items, offset, limit) {
    if (offset < 0) offset = 0;
    if (limit <= 0) limit = 100;
    let start = offset;
    let end = offset + limit;
    if (start > items.length) start = items.length;
    if (end > items.length) end = items.length;
    const out = [];
    for (let i = start; i < end; i++) {
        const it = items[i];
        if (it.file >= files.length || it.rec >= files[it.file].records.length) continue;
        const pl = files[it.file];
        const rec = pl.records[it.rec];
        const fileName = files.length > 1 ? pl.name : '';
        out.push(buildEntry(pl, rec, fileName));
    }
    return out;
}

// getTimeRange returns the first/last timestamped record time across files.
export function getTimeRange(files) {
    let first = null, last = null;
    for (const pl of files) {
        for (const r of pl.records) {
            if (!r.hasTime) continue;
            if (first === null) first = r;
            last = r;
        }
    }
    if (first === null) return {min: '', max: ''};
    const fmt = (u) => {
        const d = new Date(u * 1000);
        const p = (n) => String(n).padStart(2, '0');
        return `${d.getFullYear()}-${p(d.getMonth() + 1)}-${p(d.getDate())} ${p(d.getHours())}:${p(d.getMinutes())}:${p(d.getSeconds())}`;
    };
    return {min: fmt(first.unix), max: fmt(last.unix)};
}

// export helpers

const LABELS = {
    line: '行号', file: '文件', time: '时间', level: '级别',
    thread: '线程', logger: 'Logger', msg: '消息',
};
const PLAIN_MSG_LABEL = '内容';

function csvCell(v) {
    v = (v === null || v === undefined) ? '' : String(v);
    if (/[",\n\r]/.test(v)) {
        return '"' + v.replace(/"/g, '""') + '"';
    }
    return v;
}

// buildExportCsv builds the CSV body (without BOM) for the given visible columns.
export function buildExportCsv(cols, format, items, files) {
    const labels = {...LABELS};
    if (!format) labels.msg = PLAIN_MSG_LABEL;
    const header = cols.map((c) => labels[c] || c);
    const rows = [header];
    for (const it of items) {
        if (it.file >= files.length || it.rec >= files[it.file].records.length) continue;
        const pl = files[it.file];
        const rec = pl.records[it.rec];
        const row = [];
        for (const c of cols) {
            switch (c) {
                case 'line': row.push(String(rec.lineNo)); break;
                case 'file': row.push(pl.name); break;
                case 'time': row.push(rec.time); break;
                case 'level': row.push(rec.level); break;
                case 'thread': row.push(rec.thread); break;
                case 'logger': row.push(rec.logger); break;
                case 'msg': row.push(format ? rec.msg : readRecord(pl, rec)); break;
                default: row.push(''); break;
            }
        }
        rows.push(row);
    }
    return rows.map((r) => r.map(csvCell).join(',')).join('\r\n');
}
